source/MaterialXGenShader/Util.cpp

```diff
@@ -608,4 +608,85 @@ void findRenderableElements(const DocumentPtr& doc, std::vector<TypedElementPtr>
     }
 }
 
+bool getUIProperties(const ValueElementPtr nodeDefElement, UIProperties& uiProperties)
+{
+    if (!nodeDefElement)
+    {
+        return false;
+    }
+
+    uiProperties.uiName = nodeDefElement->getAttribute(ValueElement::UI_NAME_ATTRIBUTE);
+    uiProperties.uiFolder = nodeDefElement->getAttribute(ValueElement::UI_FOLDER_ATTRIBUTE);
+    const string& uiMinString = nodeDefElement->getAttribute(ValueElement::UI_MIN_ATTRIBUTE);
+    if (nodeDefElement->isA<Parameter>())
+    {
+        uiProperties.enumeration = nodeDefElement->getAttribute(ValueElement::ENUM_ATTRIBUTE);
+        uiProperties.enumerationValues = nodeDefElement->getAttribute(ValueElement::ENUM_VALUES_ATTRIBUTE);
+    }
+    if (!uiMinString.empty())
+    {
+        ValuePtr value = Value::createValueFromStrings(uiMinString, nodeDefElement->getType());
+        if (value)
+        {
+            uiProperties.uiMin = value;
+        }
+    }
+    const string& uiMaxString = nodeDefElement->getAttribute(ValueElement::UI_MAX_ATTRIBUTE);
+    if (!uiMaxString.empty())
+    {
+        ValuePtr value = Value::createValueFromStrings(uiMaxString, nodeDefElement->getType());
+        if (value)
+        {
+            uiProperties.uiMax = value;
+        }
+    }
+    return true;
+}
+
+bool getUIProperties(const string& path, DocumentPtr doc, const string& target, UIProperties& uiProperties)
+{
+    if (path.empty())
+    {
+        return false;
+    }
+    ElementPtr pathElement = doc->getDescendant(path);
+    if (!pathElement)
+    {
+        return false;
+    }
+    ElementPtr parent = pathElement->getParent();
+    if (!parent)
+    {
+        return false;
+    }
+
+    NodeDefPtr nodeDef = nullptr;
+    ShaderRefPtr shaderRef = parent->asA<ShaderRef>();
+    if (shaderRef)
```

Comment:
getNodeDef is kind of ugly as it's not virtual since target is not a common argument. Will need to see if it can be.